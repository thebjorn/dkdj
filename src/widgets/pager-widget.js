
// const $ = require('jquery');
// const dk = require('../boot');
// const Widget = require('../widgetcore/dk-widget.js');

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {dkconsole} from "../lifecycle/dkboot/dk-console";

export class PagerWidget extends Widget {
    constructor(...args) {
        super({
            // type: 'PagerWidget',

            pagecount: 0,               // total number of pages
            curpage: undefined,         // currently selected page (zero-based)

            classes: ['pagination'],
            css: {
                marginTop: 0,
                marginBottom: 0
            },
            template: {root: 'ul'},
            
        }, ...args);
    }
    

    /*
     *  Set the pagecount to `n`.
     */
    set_pagecount(n) {
        if (this.pagecount !== n) {
            this.pagecount = n;
            this.draw();
        }
    }

    /*
     *   Select page `n` (zero-based).
     */
    select_page(n) {
        if (n === this.curpage || this.pagecount=== 0) return;     // nothing to do
        if (n < 0 || n >= this.pagecount) {
            dkconsole.debug("PagerWidget.select_page: out of bounds ", n, this);
            return;
        }
        this.curpage = n;
        this.draw();
        this.trigger('select-page', n);
    }

    handlers() {
        const self = this;
        this.widget().on('click', 'li.page', function () {
            self.select_page(dk.$(this).data('page'));
        });
        this.widget().on('click', 'li.prev-nav', function () {
            if (self.curpage > 0) {
                self.select_page(self.curpage - 1);
            }
        });
        this.widget().on('click', 'li.next-nav', function () {
            if (self.curpage + 1 < self.pagecount) {
                self.select_page(self.curpage + 1);
            }
        });
    }
    _prerange(a, b) {
        return (b < a)? null: [a, Math.min(b + 1, a + 2)];
    }
    _postrange(a, b) {
        return (a > b)? null: [Math.max(a, b - 1), b];
    }
    draw_range(range) {
        for (let i = range[0]; i <= range[1]; i++) {
            const li = dk.$('<li/>', { page: i }).addClass('page').data('page', i-1);
            li.append(dk.$('<a href="#"/>').text(i));
            if (i === this.curpage + 1) {
                li.addClass('active');
            }
            this.widget().append(li);
        }
    }
    draw() {
        this.widget().empty();
        this.widget().append(dk.$('<li class="prev-nav"><a href="#">&laquo;</a></li>'));

        if (this.pagecount > 0) {
            let prerange = this._prerange(1, this.curpage - 3);
            let midrange = [Math.max(1, this.curpage - 3), Math.min(this.pagecount, this.curpage + 3)];
            let postrange = this._postrange(this.curpage + 3, this.pagecount);

            // console.log("PRE:", prerange);
            // console.log("MID:", midrange);
            // console.log("POST:", postrange);
            // console.log('----after...---');
            
            if (prerange && prerange[1] + 1 >= midrange[0]) {
                midrange[0] = 1;
                prerange = null;
            }
            if (postrange && midrange[1] + 1 >= postrange[0]) {
                midrange[1] = postrange[1];
                postrange = null;
            }
            // console.log("PRE:", prerange);
            // console.log("MID:", midrange);
            // console.log("POST:", postrange);
            if (prerange) {
                this.draw_range(prerange);
                this.widget().append(dk.$('<li class="ellipsis"><span>...</span></li>'));
            }
            if (midrange) {
                this.draw_range(midrange);
            }
            if (postrange) {
                this.widget().append(dk.$('<li class="ellipsis"><span>...</span></li>'));
                this.draw_range(postrange);
            }
        }

        this.widget().append(dk.$('<li class="next-nav"><a href="#">&raquo;</a></li>'));
    }

}
