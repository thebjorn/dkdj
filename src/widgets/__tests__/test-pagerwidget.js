
import $ from 'jquery';
import {PagerWidget} from "../pager-widget";
import page from "../../widgetcore/dk-page";
import utidy from "../../browser/dk-html";


test("test-pagerwidget-1", () => {
    document.body.innerHTML = `
        <div id="work">
            
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    const pw = PagerWidget.create_inside(work, {
        pagecount: 50,
        curpage: 0
    });

    expect(document.getElementById('work')).toMatchSnapshot();
    // expect(utidy(work.html())).toEqual(utidy(`
    //     <ul style="margin-top: 0px; margin-bottom: 0px;" class="PagerWidget pagination" id="pager-widget">
    //         <li class="prev-nav"><a href="#">«</a></li>
    //         <li class="page active" page="1"><a href="#">1</a></li>
    //         <li class="page" page="2"><a href="#">2</a></li>
    //         <li class="page" page="3"><a href="#">3</a></li>
    //         <li class="ellipsis"><span>...</span></li>
    //         <li class="page" page="49"><a href="#">49</a></li>
    //         <li class="page" page="50"><a href="#">50</a></li>
    //         <li class="next-nav"><a href="#">»</a></li>
    //     </ul>
    // `));
});
