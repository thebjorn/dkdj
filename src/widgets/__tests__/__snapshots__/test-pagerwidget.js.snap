// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-pagerwidget-1 1`] = `
<div
  id="work"
>
  
            
        
  <ul
    class="PagerWidget pagination"
    id="pager-widget-1"
    style="margin-top: 0px; margin-bottom: 0px;"
  >
    <li
      class="prev-nav"
    >
      <a
        href="#"
      >
        «
      </a>
    </li>
    <li
      class="page active"
      page="1"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="page"
      page="2"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="page"
      page="3"
    >
      <a
        href="#"
      >
        3
      </a>
    </li>
    <li
      class="ellipsis"
    >
      <span>
        ...
      </span>
    </li>
    <li
      class="page"
      page="49"
    >
      <a
        href="#"
      >
        49
      </a>
    </li>
    <li
      class="page"
      page="50"
    >
      <a
        href="#"
      >
        50
      </a>
    </li>
    <li
      class="next-nav"
    >
      <a
        href="#"
      >
        »
      </a>
    </li>
  </ul>
</div>
`;
